Improve wxSVGFileDC::Draw[Rotated]Text

MaartenBent opened the pull request:
Improve SVG output so text looks the same as in wxDC.

- scale text to screen DPI
- new tag for preserving white-space
- draw background rectangles per line, so background colour of shorter lines is only behind the text
- add support for fractional point size from wxFont

MaartenBent committed:
Implement GetContentScaleFactor for wxMSWDCImpl

MaartenBent committed:
Refactor wxSVGFileDCImpl::DoDrawRotatedText

Create the style string outside the loop, it can determined once and used for
each line.

MaartenBent committed:
Add replacement for deprecated white-space tag in wxSVGFileDC output

MaartenBent committed:
Fix wxSVGFileDC text size when using non-default screen DPI

The text in a wxDC is adjusted to the screen DPI, but text in a SVG file is DPI
independent. Scale the text attribute so it has the same size as in the wxDC.

Scaling will modify the position of the attribute, correct this with translate.

MaartenBent committed:
Improve wxSVGFileDC text background colour

Draw the background rectangle per line, instead of one big rectangle.

When there are multiple lines and one line is shorter, the area behind that
line should not get a background colour.

MaartenBent committed:
Use fractional point size for font in wxSVGFileDC output

vadz commented:
Shouldn't we allow specifying the DPI to use somehow here? Taking it from the primary monitor risks not always being the right thing to do (even if it's clearly better than the current situation).

MaartenBent commented:
Specifying only a DPI doesn't help. `wxScreenDC` is used for `GetTextExtent`, and (as far as I know) there is no way of getting text size for a specific DPI (in win32 API).
And the goal is to make the SVG look exactly the same as `wxDC`. by using the primary monitor, the text size will always be the same as the `wxDC` output.

In the pm-dpi branch I added `SetWindow` to `wxDC`. This would allow to use the associated `wxWindow` (at any DPI) to get the text size instead of using `wxScreenDC`.

vadz commented:
Just to be clear, by "specifying the DPI somehow", I didn't mean adding `double dpi` parameter but rather allowing to specify the DC/window to use. But after thinking more about it, I realized that this didn't actually make any sense -- SVG output shouldn't depend on the screen DPI, but neither should it depend on any window at all.

And after looking at the code, everything is already alright because the `wxScreenDC` used here is consistent with `wxScreenDC` used in `wxSVGFileDCImpl::DoGetTextExtent()`, so this PR is correct and I can merge it right now if you'd like.

However I now wonder about another possible improvement: shouldn't this
scaling be actually done inside `wxSVGFileDCImpl::DoGetTextExtent()` itself
(and also in `GetChar{Height,Width}()`, of course)? It would seem to be more
logical to return DPI-independent values from there, wouldn't it?

MaartenBent commented:
> However I now wonder about another possible improvement: shouldn't this scaling be actually done inside `wxSVGFileDCImpl::DoGetTextExtent()` itself (and also in `GetChar{Height,Width}()`, of course)? It would seem to be more logical to return DPI-independent values from there, wouldn't it?

The problem is that the `<text />` attribute in the SVG will always have the same size, it is not related to 
`DoGetTextExtent` in any way. Changing the returned values does not change the size of the text. It can only be scaled when creating the attribute.

vadz commented:
I'm not sure I understand this... The PR generates `textlength=${ww/scale}` where `ww` is the width returned by `GetTextExtent()`. If `wxSVGFileDCImpl::DoGetTextExtent()` already did this scaling internally, you could just use `ww` without any scaling in `DoDrawRotatedText()`, couldn't you?

Are you worried about rounding errors or is there something else?

MaartenBent commented:
`textlength` is affected by `transform="scale(x)"`. I divide it first so after the transform it will be `ww` again.

vadz commented:
Oh, I missed that, thanks. But this means that the generated SVG depends on the DPI of the system it was created on, which definitely seems to be wrong, doesn't it?

Could we avoid having to use the `scale` transform by scaling the coordinates/extents instead?

MaartenBent commented:
The generated SVG looks the same as the `wxDC` it was generated from. If the `wxDC` is on a high-dpi screen then it will have bigger text, and so will the SVG.

See these two examples from the svg sample (change extension to svg):
[Text 96.txt](https://github.com/wxWidgets/wxWidgets/files/3468916/Text.96.txt)
[Text 192.txt](https://github.com/wxWidgets/wxWidgets/files/3468917/Text.192.txt)

There is only an x and y coordinate where to place the text. And `textLength` for how wide the text should be. None of these change the actual size.

vadz commented:
> The generated SVG looks the same as the `wxDC` it was generated from. If the `wxDC` is on a high-dpi screen then it will have bigger text, and so will the SVG.

Sorry, I must be missing something really fundamental here, but what do you mean by "generated from"? `wxSVGFileDC` can be created without any reference to any other `wxDC` and I don't understand at all why would it be desirable to produce different SVG output depending on the DPI.

IMO we should always generate the same output, no matter what.

Shouldn't we?

MaartenBent commented:
I use the svg sample and drawing sample to verify the svg output. The samples draw something on a `wxDC` and then I save that to SVG. I try to make them look identical. At higher DPI the text in `wxDC` gets bigger, so I did the same for SVG (to fix the issue reported in #1275).

Your suggestion is a different approach. Instead of making the text bigger, make the result of `GetTextExtent` smaller so `textLength` will be correct and the text does not look stretched out.

It is a choice between _'all DCs should look the same'_ and _'svg output should always be the same'_. If you want the latter, I can try to implement that.

vadz commented:
I really think that "SVG output should always be the same" is the "obvious" answer. At least I never even considered the possibility that we could intentionally do something different...

IMO SVG, PostScript, PDF etc are platform-independent formats and the results shouldn't depend on where exactly they were generated. Of course, they're vector formats too, but IMO this is not really different from e.g. PNG: it would be surprising and wrong if the output `.png` file depended on whether it was created under MSW with normal DPI or under Mac with high DPI. And the same should be true here.

MaartenBent committed:
Use DPI independent text size in wxSVGFileDC

Don't let the text size in the SVG file depend on the DPI of the system it was
created on. Let GetTextExtent, GetCharHeight and GetCharWidth return sizes
based on the default (96) DPI.

MaartenBent commented:
I added a commit to fix the DPI issue, so if you agree this can be merged. I'd like the do my other `wxSVGFileDC` changes on top of this one.

vadz committed:
Minor formatting fix

vadz committed:
Avoid calling wxScreenDC::SetFont() more than once

This seems unnecessary when we can call it only once even with high DPI
screens.

vadz commented:
Could we avoid an extra `SetFont()` call with c91c94a?

Please let me know if I'm missing anything. If not, I'll merge it soon. TIA!

MaartenBent commented:
Look good.
